asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py

```diff
@@ -47,18 +69,57 @@ def sjoin_df_epc_with_listed_buildings(
     )
 
     dfs = []
-
     # Group based on the created index
     data_partitioned = epc_df.with_columns(partitions).partition_by("chunk_id")
     logging.info(f"Adding listed buildings to EPC in {len(data_partitioned)} chunks")
     for epc_chunk in tqdm(data_partitioned):
-        epc_gdf = lat_lon.generate_gdf_uprn_coords(df=epc_chunk)[["UPRN", "geometry"]]
-        df = epc_gdf.sjoin(listed_buildings_gdf, how="inner", predicate="intersects")[
-            ["UPRN", "listed_building_grade"]
-        ].drop_duplicates(subset="UPRN")
-
+        df = sjoin_df_epc_listed_buildings(epc_chunk, listed_buildings_gdf)
         dfs.append(df)
 
     df = pl.from_pandas(pd.concat(dfs))
 
+    return df.select(["UPRN", "listed_building"])
+
+
+def sjoin_df_epc_listed_buildings(
+    epc_df: pl.DataFrame, listed_buildings_gdf: gpd.GeoDataFrame, distance: float = 5
+) -> pd.DataFrame:
+    """
+    Spatial join EPC UPRNs with listed buildings using `geopandas.GeoDataFrame.sjoin_nearest` where Point or MultiPoint
+    geometries detected, and `geopandas.GeoDataFrame.sjoin` where Polygons or MultiPolygons detected.
+
+    Args:
+        epc_df (pl.DataFrame): EPC dataset with X and Y coordinates per UPRN
+        listed_buildings_gdf (gpd.GeoDataFrame): listed buildings data
+        distance (float): maximum distance (m) within which to query for nearest geometry where `sjoin_nearest` used.
+                          Must be greater than 0. Default 5.
+
+    Returns:
+        pd.DataFrame: EPC UPRNs in listed buildings
+    """
+    epc_gdf = lat_lon.generate_gdf_uprn_coords(df=epc_df, usecols=["UPRN"])
+    if any(
+        [
+            expr in listed_buildings_gdf.geom_type.unique()
+            for expr in ["Point", "MultiPoint"]
+        ]
+    ):
+        df = epc_gdf.sjoin_nearest(
+            listed_buildings_gdf, how="inner", max_distance=distance
+        )[["UPRN", "listed_building"]].drop_duplicates(subset="UPRN")
```

Comment:
Yes that's correct. So this should return a df with UPRNs and then a boolean flag for being in a listed building. Because it's an inner join, this will only return UPRNs which are in listed buildings.